Add content for FedCM extensions (#28585)
mdn/content: files/en-us/web/api/fedcm_api/index.md

{{SeeCompatTable}}{{DefaultAPISidebar("FedCM API")}}

> The **Federated Credential Management API** (or _FedCM API_) provides a standard mechanism for identity providers (IdPs) to make identity federation services available on the web in a privacy-preserving way, without the need for third-party cookies and redirects. This includes a JavaScript API enabling the use of federated authentication for activities such as signing in or signing up on a website.

Reviewer: ```suggestion
The **Federated Credential Management API** (or _FedCM API_) provides a standard mechanism for identity providers (IdPs) to make identity federation services available on the web in a privacy-preserving way, without the need for third-party cookies and redirects. This includes a JavaScript API that enables the use of federated authentication for activities such as signing in or signing up on a website.
```

  - : The URL of the user's avatar image.
- `approved_clients` {{optional_inline}}
  - : An array of RP clients that the user has registered with.

> - `login_hints` {{optional_inline}}

Author: Thanks. I've tweaked the wording in the property description to make this clearer.

## RP sign-in using the JavaScript API

RPs can call {{domxref("CredentialsContainer.get", "navigator.credentials.get()")}} with an `identity` option to make a request for users to sign in to the RP with the IdP, using the available IdP config and endpoints (as described above). A typical request might look like this:

```js
async function signIn() {
  const identityCredential = await navigator.credentials.get({
    identity: {
      context: "signup",
      providers: [
        {
          configURL: "https://accounts.idp.example/config.json",
          clientId: "********",
          nonce: "******",
          loginHint: "user1@example.com",
        },
      ],
    },
  });
}
```

The `identity.providers` property takes an array of objects containing the path to an IdP config file and the RP's client identifier issued by the IdP; the example above also includes a couple of optional features:

> - `identity.context` specifies the context in which the user is authenticating with FedCM (for example is it a first-time signup for this account, or a sign-in with an existing account?). The browser uses this to vary the text in its FedCM sign-in UI so that it better suits the context.

Author: Done

- `identity.providers.nonce` provides a random nonce value that ensures the response is issued for this specific request, preventing {{glossary("replay attack", "replay attacks")}}.

> - `identity.providers.loginHint` provides a hint about which account option(s) the browser should present for user sign-in. This hint is matched against the `login_hints` values that the IdP provides from the [accounts list endpoint](#the_accounts_list_endpoint).

Reviewer: ```suggestion
- `identity.providers.loginHint` provides a hint about the account option(s) the browser should present for user sign-in. This hint is matched against the `login_hints` values that the IdP provides from the [accounts list endpoint](#the_accounts_list_endpoint).
```

> - `identity.context` specifies the context in which the user is authenticating with FedCM (for example is it a first-time signup for this account, or a sign-in with an existing account?). The browser uses this to vary the text in its FedCM UI so that it better suits the context.
> - `identity.providers.nonce` provides a random nonce value that ensures the response is issued for this specific request, preventing {{glossary("replay attack", "replay attacks")}}.
> - `identity.providers.loginHint` provides a hint as to which account option(s) the browser should provide for the user to sign in with, which is matched against the `login_hints` values provided by the IdP from the [accounts list endpoint](#the_accounts_list_endpoint).

Reviewer: ```suggestion
- `identity.context` specifies the context in which the user is authenticating with FedCM. For example, is it a first-time signup for this account, or a sign-in with an existing account?. The browser uses this information to vary the text in its FedCM UI to better suit the context.
- `identity.providers.nonce` provides a random nonce value that ensures the response is issued for this specific request, preventing {{glossary("replay attack", "replay attacks")}}.
- `identity.providers.loginHint` provides a hint about which account option(s) the browser should present for user sign-in. This hint is matched against the `login_hints` values that the IdP provides from the [accounts list endpoint](#the_accounts_list_endpoint).
```

The availability of FedCM within `<iframe>`s enables a couple of use cases:

- Larger sites won't want a third-party sign-in script to gain control over the top-level frame; instead they will want to add that script and invoke FedCM from within an {{htmlelement("iframe")}}.
- Some `<iframes>` may themselves require federated authentication.

## Interfaces

- {{domxref("IdentityCredential")}}
  - : Represents a user identity credential arising from successful federated authentication. A successful {{domxref("CredentialsContainer.get", "navigator.credentials.get()")}} call that includes an `identity` option fulfills with an {{domxref("IdentityCredential")}} instance.
- {{domxref("IdentityProvider")}}

>   - : Represents an IdP and provides access to related information. The static {{domxref("IdentityProvider.getUserInfo_static", "IdentityProvider.getUserInfo()")}} method returns information about a previously signed in user on their return to an IdP, which can be used to provide a personalized welcome message and sign-in button.

Reviewer: ```suggestion
  - : Represents an IdP and provides access to related information. The static {{domxref("IdentityProvider.getUserInfo_static", "IdentityProvider.getUserInfo()")}} method returns information about a user who was previously signed in, when they return to an IdP. This information can be used to display a personalized welcome message and sign-in button.
```

## Extensions to other interfaces

- {{domxref("CredentialsContainer.get()")}}, the `identity` option.

>   - : `identity` is an object containing details of federated IdPs that a relying party (RP) website can use to sign users in. Causes a `get()` call to initiate a request for a user to sign in to a RP with an IdP.

Reviewer: ```suggestion
  - : `identity` is an object containing details of federated IdPs that a relying party (RP) website can use to sign users in. It causes a `get()` call to initiate a request for a user to sign in to an RP with an IdP.
```
